### ISSUE TITLE
Question about alignment with riscv-org repository

### ISSUE THREAD

**Issue by natirapaport** [riscv/riscv-openocd#790]
Can anyone tell how much riscv-openocd repo, is aligned with official [openocd-org/openocd](https://github.com/openocd-org/openocd) repository?
The latest riscv-openocd release is from Dec-22, 2018 , but on the other hand the version of riscv-openocd is 0.11 which is very close to openocd-org/openocd latest version (0.12) .
From here, I couldn’t tell where is the latest pull-request from riscv-openocd fork to openocd-org/openocd repo. Can someone please point at that commit?
Thanks,
Nati

**Comment by TommyMurphyTM1234** [comment]
For what it's worth, I think that this is the last time that this repo was sync'd with the upstream master upstream OpenOCD.

* https://github.com/riscv/riscv-openocd/pull/696

This happens every so often and changes here are upstreamef periodically too.

You could also always clone both repos and diff them or use git to check for differences.

**Comment by timsifive** [comment]
I've been very behind on this. I actually am working on another small merge that I thought would be trivial, but then there was a regression (or possibly a bug I hadn't caught before).

As for the common ancestor:
```
$ git merge-base master riscv
66335683fec62ac89da48d64932fd9d082314225
```

**Comment by oferShinaar** [comment]
@timsifive this repo is contributing back to the openocd repo? if so, can you share the schedule points?
On the other hand, if all riscv-openocd is in this repo what is the solid-working branch (no tags or releases I saw)?

thanks

**Comment by timsifive** [comment]
> @timsifive this repo is contributing back to the openocd repo? if so, can you share the schedule points?

Yes, very slowly, when I have time. I'd love help with this. There is no schedule.

> On the other hand, if all riscv-openocd is in this repo what is the solid-working branch (no tags or releases I saw)?

master in this branch should always be working well. There are no official tags or releases.

**Comment by natirapaport** [comment]
@timsifive - do you know to tell if riscv-openocd pass the same , or similar testing procedures like the openocd-org one? In other words, how much it is considered to be reliable comparing to openocd-org?

**Comment by timsifive** [comment]
I'm not aware of any systematic testing done by upstream OpenOCD. Since riscv-openocd is behind mainline, there are almost certainly bugfixes there that are not here.

**Comment by natirapaport** [comment]
@timsifive , Thank you for all the information.
Another question please - how stable is riscv-openocd for working with multi-hart core? Is there any commit that known to be stable? I'm asking because we encounter various issues, suspected to be related to openocd, when working with 2 threads (with the tip commit from this repo)

**Comment by timsifive** [comment]
Multi-hart cores is a very common use case, so I would expect it to work well. If there are problems, feel free to open an issue and include the `openocd -d` log.

**Comment by TommyMurphyTM1234** [comment]
Just noting this PR here for reference purposes:

* https://github.com/riscv/riscv-openocd/pull/816

**Comment by TommyMurphyTM1234** [comment]
Hi @timsifive - quick question which, offhand, I don't see mentioned elsewhere.... Do you have any feel for how far off the day might be that further RISC-V development/enhancement can be done directly upstream obviating the need for a RISC-V fork of OpenOCD? Or, maybe put another way, what are the main big tasks outstanding before this might be feasible? Thanks.

**Comment by timsifive** [comment]
At the current pace, this will never happen.

The branches are drifting further apart, and I don't have the bandwidth to bring them together. The process to bring them together is pretty straightforward:
1. Merge changes from upstream into this branch.
2. Merge changes from this branch into upstream.

**Comment by TommyMurphyTM1234** [comment]
Thanks @timsifive. I fully understand that you (and others who contribute) have limited bandwidth and/or too much to do.

Just for my own edification, is there anywhere any outline of stuff that remains to be implemented? Including anything that arises from further revisions of the debug (and trace?!) spec(s)? 

Apologies for all the questions. I hope that they're ok and that this is an appropriate forum?

**Comment by timsifive** [comment]
There is no todo list. OpenOCD development tends to be done as-needed. Lately the amount of work done in this branch is actually increasing as more people are using RISC-V chips and OpenOCD. I think it's unlikely this pace will slow in the foreseeable future.

**Comment by TommyMurphyTM1234** [comment]
> There is no todo list. OpenOCD development tends to be done as-needed. Lately the amount of work done in this branch is actually increasing as more people are using RISC-V chips and OpenOCD. I think it's unlikely this pace will slow in the foreseeable future.

Thanks @timsifive - that's fine. I was just wondering if there was any sort of outline of a schedule/task list. No worries if there's not.

**Comment by aap-sc** [comment]
Hi @timsifive ,  would you kindly clarify on this one:

> The branches are drifting further apart, and I don't have the bandwidth to bring them together. The process to bring them together is pretty straightforward:
> Merge changes from upstream into this branch.
> Merge changes from this branch into upstream.

Is the issue only with this process (likely, both merges require significant amount of time)? Or there are some other issues preventing riscv-openocd development to move to the mainline?

**Comment by timsifive** [comment]
The issue is only with the process. There are no legal or technical reasons the branches cannot be brought together.

**Comment by zqb-all** [comment]
> Merge changes from upstream into this branch.
> Merge changes from this branch into upstream.

If someone helps with this, what tests should be done after the merge to make sure the merge didn't introduce any problems？

**Comment by timsifive** [comment]
> If someone helps with this, what tests should be done after the merge to make sure the merge didn't introduce any problems?

I rely on https://github.com/riscv-software-src/riscv-tests/tree/master/debug. I run `make all` there. Then I also run the tests against HiFive Unleashed, and if riscv-011 has changed, HiFive1 as well. For the hardware platforms some of the tests fail. Unfortunately that's not documented anywhere.

**Comment by mrv96** [comment]
I'm waiting too about upstream merging. Is someone working on this?

**Comment by TommyMurphyTM1234** [comment]
> I'm waiting too about upstream merging. Is someone working on this?

Waiting on others (in particular @timsifive primarily) to do all the work isn't exactly in the spirit of open source... 😐

But, anyway...

* https://github.com/riscv/riscv-openocd/pull/844

**Comment by mrv96** [comment]
Thank you for the quick reply.

I'm already a contributor of this repository... however upstream merge is quite tricky and potentially harmful, hence is better if done by a maintainer of the repository.

Despite this, I was willing to try by myself in case of negative answer to my question

**Comment by timsifive** [comment]
I'm working on this here and there. My general process is:
1. Merge down changes from upstream. We are considerably behind. #844 would make us less behind. It's waiting https://github.com/riscv-software-src/riscv-tests/pull/469 before we can merge it.
2. Upstream changes.

We don't *have* to do things in that order, and any upstream change (however small) the reduces the diff between the two branches would be helpful.
